docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb

```diff
@@ -0,0 +1,681 @@
+    "\n",
+    "# save the model\n",
+    "simple_model_name = 'simple.onnx'\n",
+    "onnx.save(model, simple_model_name)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "When loading the model we created above, Nuphar execution provider would be used. We can use settings string to check the generated code.\n",
+    "\n",
+    "Because of redirection of output, we dump the lowered code from a subprocess to a log file:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "import subprocess\n",
+    "import sys\n",
+    "\n",
+    "code_to_run = '''\n",
+    "import onnxruntime\n",
+    "s = 'codegen_dump_lower:verbose'\n",
+    "onnxruntime.capi._pybind_state.set_nuphar_settings(s)\n",
+    "sess = onnxruntime.InferenceSession('simple.onnx')\n",
+    "'''\n",
+    "\n",
+    "log_file = 'simple_lower.log' \n",
+    "with open(log_file, \"w\") as f:\n",
+    "  subprocess.run([sys.executable, '-c', code_to_run], stdout=f, stderr=f)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The lowered log is similar to C source code, but the whole file is lengthy to show here. Let's just check the last few lines that are most important:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "['produce node4 {\\n',\n",
+       " '  for (ax0, 0, seq) {\\n',\n",
+       " '    for (ax1, 0, batch) {\\n',\n",
+       " '      for (ax2.outer, 0, 64) {\\n',\n",
+       " '        node4[ramp((((((ax0*batch) + ax1)*64) + ax2.outer)*16), 1, 16)] = (input[ramp((((((ax0*batch) + ax1)*64) + ax2.outer)*16), 1, 16)] + (input[ramp((((((ax0*batch) + ax1)*64) + ax2.outer)*16), 1, 16)]*(input[ramp((((((ax0*batch) + ax1)*64) + ax2.outer)*16), 1, 16)] + (input[ramp((((((ax0*batch) + ax1)*64) + ax2.outer)*16), 1, 16)]*(input[ramp((((((ax0*batch) + ax1)*64) + ax2.outer)*16), 1, 16)] + input[ramp((((((ax0*batch) + ax1)*64) + ax2.outer)*16), 1, 16)])))))\\n',\n",
+       " '      }\\n',\n",
+       " '    }\\n',\n",
+       " '  }\\n',\n",
+       " '}\\n',\n",
+       " '\\n']"
+      ]
+     },
+     "execution_count": 3,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "with open(log_file) as f:\n",
+    "    log_lines = f.readlines()\n",
+    "\n",
+    "log_lines[-10:]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The compiled code shows that the nodes of Add/Mul is fused in a single function, and vectorization is applied in the loop. This fusion happened inside compiler used by Nuphar execution provider, and it does not require model editing.\n",
+    "\n",
+    "Next, let's run inference on the model and compare with numpy in accuracy and speed:"
```

Comment:
probably modify the sentence to "and compare the accuracy and performance with numpy"